### DISCUSSION TITLE
Implementing OpenCL backend

### DISCUSSION THREAD

**Discussion by artyom-beilis** [microsoft/onnxruntime#18540] [General]
Hello,

I'm developer of the [dlprimitives](https://github.com/artyom-beilis/dlprimitives) - opencl implementation of neural network. It also has some basic ONNX support and I tested it on many detection networks.

However implementing full implementation of ONNX backend is rather complicated. I've seen that it is possible to integrate with ONNXRuntime and I wonder what benefits I can have.

There are lots of operators to implement and I wonder if ONNX Runtime can help me with

1. Memory and graph management - i.e. allocating of the tensors and intermediate data - optimising graph memory use/memory reuse?
2. Implementing of some operators in terms of other? For example if I implement gemm automatically of matmul exists and other way around? Implement norm layer in terms of other operators etc.
3. General graph operators that mostly handle tensor shapes
4. Implementing "higher level" opsets in terms of lower opsets?

If so is there a "minimal" set of operators to implement to get it running?

**Comment by jdemirff** [comment]
Is there any update about that ?

**Comment by artyom-beilis** [comment]
There is a very basic implementation of ONNX as part of dlprimitives (works for major image classification networks like ResNet50) but not as part of ONNXRuntime

So I hadn't started looking into it yet due to lack of time and resources
